### PR TITLE
DOC: Prepend `ma.` to references in ``numpy.ma``

### PR DESCRIPTION
**Pull request by takanori-pskq**
Related #13114 : separated from #16300 for further discussion

### CODE REVIEW

**Comment by takanori-pskq** [comment]
> > Instead of just copying docstrings from numpy versions, how about writing the docstrings of the functions in numpy.ma like "This function is numpy.ma version of the function foo. See numpy.foo for details"?
> 
> I tend to agree in principle, but I don't think that this is a general solution. While some functions in `ma` simply use the docstring from their `np` counterpart , some don't. There are examples of both patterns (and others) in `numpy/ma/core.py` for example. I am not familiar enough with `np.ma` to say whether an approach like this would scale well.

(from the conversation in #16300 )

It's enough to replace to "This function is ..." only when the function simply copies the docstring in `np` version. If the function doesn't copy (the function has proper docstring apart from the `np` version), we don't have to change.

File: numpy/ma/core.py [hunk 1]
```diff
@@ -5605,7 +5605,7 @@ def sort(self, axis=-1, kind=None, order=None,
         --------
         numpy.ndarray.sort : Method to sort an array in-place.
         argsort : Indirect sort.
```
Comment (Qiyu8) [inline numpy/ma/core.py:5607]: should add prefix too.

File: numpy/ma/core.py [hunk 2]
```diff
@@ -5605,7 +5605,7 @@ def sort(self, axis=-1, kind=None, order=None,
...
-        lexsort : Indirect stable sort on multiple keys.
+        numpy.lexsort : Indirect stable sort on multiple keys.
         searchsorted : Find elements in a sorted array.
```
Comment (Qiyu8) [inline numpy/ma/core.py:5609]: should add prefix too.

File: numpy/core/fromnumeric.py [hunk 3]
```diff
@@ -3148,8 +3148,8 @@ def size(a, axis=None):
     See Also
     --------
     shape : dimensions of array
```
Comment (Qiyu8) [inline numpy/core/fromnumeric.py:3150]: should be numpy.ma.shape.

Comment (rossbar) [reply numpy/core/fromnumeric.py:3150]: > should be numpy.ma.shape.

This is the docstring for `np.size`, so it should be referencing functions from the `numpy` namespace, not `np.ma` I think.

**Review by rossbar — COMMENTED** [review]
@Qiyu8 the reason I suggested @takanori-pskq split this off from #16300 was to see if we could find a way to make the links work in `np.ma` *without* having to prepend `numpy` to everything, so as not to clutter up the docstring for the numpy functions.

Perhaps it's worth looking at the `getdoc` method of the `_frommethod` and/or `_convert2ma` classes? Again, I'm not too familiar with the details of how `np.ma` is documented, so please share any ideas. I just think it's worth exploring if we can fix some of these broken links in `np.ma` without having to touch the docstrings for the corresponding `np` objects.

**Comment by takanori-pskq** [comment]
Some functions use neither `_frommethod` nor `_convert2ma`, and just copy (`foo.__doc__ = np.foo.__doc__`). If `_convert2ma` can be applied for these functions, it is enough to see the `getdoc` method in `_convert2ma`, and add the process to prepend `numpy.` to the links in the docstring which start other than `numpy.`.

**Comment by rossbar** [comment]
> If _convert2ma can be applied for these functions, it is enough to see the getdoc method in _convert2ma, and add the process to prepend numpy. to the links in the docstring which start other than numpy.

This would work, though then you'd be automatically overriding any references to `np.ma` versions of functions when they're defined. Take `np.ma.var` for instance: there are six references in the See Also section, five of which are references to functions. Of those, two (`std` and `mean`) have `np.ma` counterparts, whereas the other three do not. If you prepend `numpy` to everything, than you lose the specific references to the `ma` versions of `mean` and `std`. Since the docstrings appear to be simply copied over from the non-masked versions, it doesn't seem like you'd lose anything, but in the case where the `ma` version of a function has an `ma`-specific docstring, you'd be overriding the reference to it.

**Comment by mattip** [comment]
What is the status of this experiment?

**Review by rossbar — COMMENTED** [review]
Thanks for the ping @mattip 

tl;dr - upside: fixes some broken links | downside: makes some `See Also` sections uglier and in some cases redirects away from `ma` functions when they would be a better fit.

I'm still ambivalent about this. See the proposed changes to the `See Also` section of `var`. Currently, the `ma.var` docs link to the `ma` versions of `std` and `mean` in the `See Also` section: https://numpy.org/devdocs/reference/generated/numpy.ma.var.html. With this change, those links will instead point to the `numpy` counterparts, rather than `numpy.ma`. This also clutters-up the `See Also` sections in the core functions, as prepending `numpy.` is not necessary to make those links work.

The real problem stems from how `ma` docs are generated. Ideally there would be a solution that would add e.g. `.. currentmodule`'s etc. where appropriate when the `ma` docs are being generated, but there doesn't seem to be a one-size-fits-all solution. 

There seems to be a nice push to get #13114 closed (big +1 IMO), so I'm okay accepting this as-is to further that goal and revisiting to fix the details later on.

**Comment by mattip** [comment]
> Ideally there would be a solution that would add e.g. `.. currentmodule`'s etc. where appropriate when the ma docs are being generated

Could this be done with a custom template option in the [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html) in [doc/source/referencemaskedarray.baseclass.rst](https://github.com/numpy/numpy/blob/v1.19.2/doc/source/reference/maskedarray.baseclass.rst)?

**Comment by takanori-pskq** [comment]
I'll try using template.

**Comment by takanori-pskq** [comment]
Now I work to fix this problem in #17515 .

**Comment by takanori-pskq** [comment]
Modified to prepend `ma.` instead of `numpy.`. (see https://github.com/numpy/numpy/pull/17515#issue-500450253).

File: numpy/ma/core.py [hunk 4]
```diff
@@ -3538,11 +3538,12 @@ def harden_mask(self):
         Force the mask to hard.
 
         Whether the mask of a masked array is hard or soft is determined by
-        its `hardmask` property. `harden_mask` sets `hardmask` to True.
+        its `~ma.MaskedArray.hardmask` property. `harden_mask` sets
+        `~ma.MaskedArray.hardmask` to ``True``.
```
Comment (mattip) [inline numpy/ma/core.py:3542]: This works nicely

**Review by mattip — COMMENTED** [review]
Looks good, the template works as expected.

File: numpy/ma/core.py [hunk 5]
```diff
@@ -1489,7 +1489,7 @@ def is_mask(m):
 
     See Also
     --------
-    isMaskedArray : Test whether input is an instance of MaskedArray.
+    ma.isMaskedArray : Test whether input is an instance of MaskedArray.
```
[the remark this answers is not in the document]
Comment (takanori-pskq) [reply numpy/ma/core.py:1492]: `is_mask` is on the page [`routines.ma`](https://16783-908607-gh.circle-artifacts.com/0/doc/build/html/reference/routines.ma.html) (I think this is because `is_mask` is a function in `np.ma`, not a method of the class `MaskedArray`). And we might have to insert `isMaskedArray` (also `isarray`, `isMA`) right after `is_mask`.

**Comment by mattip** [comment]
Thanks @takanori-pskq.
